nat-lab/tests/utils/batching.py

```diff
@@ -0,0 +1,161 @@
+
+        sorted_values = sorted(self._values)
+        max_val = sorted_values[-1]
+
+        if max_val >= buckets * bucket_size:
+            raise ValueError(
+                f"Histogram doesn't fit the data({max_val}). The max value is {max_val} but the histogram has {buckets} buckets with a width of"
+                f" {bucket_size}, the maximum value it can fit is {buckets * bucket_size}"
+            )
+
+        hs = [0] * buckets
+
+        for v in sorted_values:
+            bucket_index = int(v / bucket_size)
+            hs[bucket_index] += 1
+
+        return hs
+
+    def __repr__(self) -> str:
+        return f"Histogram of {len(self._values)} values"
+
+
+async def capture_histograms(
+    container_name: str, duration_s: int, filters: list["PacketFilter"]
+) -> List[Tuple[str, List[int]]]:
+    cmd_rm = f"docker exec --privileged {container_name} rm /home/capture.pcap"
+    os.system(cmd_rm)
+
+    iface = "any"
+    capture_path = "/home/capture.pcap"
+
+    cmd = f"docker exec -d --privileged {container_name} tcpdump -i {iface} -U -w {capture_path}"
+    res = os.system(cmd)
+    if res != 0:
+        raise RuntimeError(f"Failed to launch tcpdump on {container_name}")
+
+    await asyncio.sleep(duration_s)
+
+    local_path = f"{container_name}.pcap"
+    subprocess.run([
+        "docker",
+        "cp",
+        container_name + ":" + "/home/capture.pcap",
+        local_path,
+    ])
+
+    histograms = []
+    for f in filters:
+        # add extra time for histogram as tcpdump was not killed immediately
+        hs = _generate_histogram(local_path, duration_s + 1, 1, f)
+        histograms.append((f._name, hs))
+
+    os.unlink(local_path)
+
+    cmd_rm = f"docker exec --privileged {container_name} pkill tcpdump"
+    os.system(cmd_rm)
+
+    return histograms
+
+
+# Render ASCII histogram drawing for visual inspection
+def draw_histogram(name: str, data: List[int], max_height=None):
+    if not data:
+        print(f"No data provided for {name}")
+        return
+
+    max_value = max(data)
+
+    if max_height is None:
+        max_height = max_value
+
+    scaled_data = [math.ceil((value / max_value) * max_height) for value in data]
+    for row in range(max_height, 0, -1):
+        line = ""
+        for value in scaled_data:
+            if value >= row:
+                line += "█"
+            else:
+                line += " "
+        line = "|" + line
+        print(line)
+
+    print(f"+{'-' * (len(data))}")
+    print(f"0{' ' * (len(data)-1)}{len(data)}")
+    print(f"^-Histogram of {name}")
+
+
+class PacketFilter:
+    def __init__(self, name: str, filter: Callable[[Any], bool]):
+        self._name = name
+        self._filter = filter
+
+    def filter(self, packet) -> bool:
+        return self._filter(packet)
+
+    def __str__(self):
+        return f"PacketFilter({self._name})"
+
+
+def _generate_histogram(
+    pcap_path: str, buckets: int, bucket_width: int, filter: PacketFilter
```

Comment:
Hmm, what is the purpose of wrapping this `Callable` with `PacketFilter`?

Reply:
No big reason except `PacketFilter` also has a `_name` property to it which I find nicer to use instead of passing a tuple of `(str, Callable)`

Reply:
So I actually just replaced it with `Callable` and that went _fine_. It definitely fixed a funny looking part of `filter.filter(pkt)`.